File: android/src/main/java/expo/modules/nearbyconnections/ExpoNearbyConnectionsModule.kt
package expo.modules.nearbyconnections

import android.content.Context
import androidx.core.os.bundleOf
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException

class ExpoNearbyConnectionsModule : Module() {
    override fun definition() = ModuleDefinition {
        Events(*EventName.entries.map { it.toString() }.toTypedArray())

        Name(MODULE_NAME)

        AsyncFunction("isPlayServicesAvailable") {
            val result = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)
            val isAvailable = result == ConnectionResult.SUCCESS

            return@AsyncFunction isAvailable
        }

        AsyncFunction("startAdvertise") { name: String, strategy: Double?, promise: Promise ->
            nearbyConnectionsModule.startAdvertise(name, strategy).addOnSuccessListener { result ->
                promise.resolve(result)
            }.addOnFailureListener {
                promise.reject(
                    CodedException(
                        "E_NEARBY_CONNECTIONS_START_ADVERTISE_FAILED",
                        "Failed to start advertising",
                        it
                    )
                )
            }
        }

        AsyncFunction("stopAdvertise") {
            return@AsyncFunction nearbyConnectionsModule.stopAdvertise()
        }

        AsyncFunction("startDiscovery") { name: String, strategy: Double?, promise: Promise ->
            nearbyConnectionsModule.startDiscovery(name, strategy).addOnSuccessListener { result ->
                promise.resolve(result)
            }.addOnFailureListener {
                promise.reject(
                    CodedException(
                        "E_NEARBY_CONNECTIONS_START_DISCOVERY_FAILED",
                        "Failed to start discovery",
                        it
                    )
                )
            }
        }

        AsyncFunction("stopDiscovery") {
            return@AsyncFunction nearbyConnectionsModule.stopDiscovery()
        }

        AsyncFunction("requestConnection") { advertisePeerId: String, promise: Promise ->
            nearbyConnectionsModule.requestConnection(advertisePeerId).addOnSuccessListener {
                promise.resolve(null)
            }.addOnFailureListener {
                promise.reject(
                    CodedException(
                        "E_NEARBY_CONNECTIONS_REQUEST_CONNECTION_FAILED",
                        "Failed to request connection",
                        it
                    )
                )

            }
        }

        AsyncFunction("acceptConnection") { targetPeerId: String, promise: Promise ->
            nearbyConnectionsModule.acceptConnection(targetPeerId).addOnSuccessListener {
                promise.resolve(null)
            }.addOnFailureListener {
                promise.reject(
                    CodedException(
                        "E_NEARBY_CONNECTIONS_ACCEPT_CONNECTION_FAILED",
                        "Failed to accept connection",
                        it
                    )
                )
            }
        }

        AsyncFunction("rejectConnection") { targetPeerId: String, promise: Promise ->
            nearbyConnectionsModule.rejectConnection(targetPeerId).addOnSuccessListener {
                promise.resolve(null)
            }.addOnFailureListener {
                promise.reject(
                    CodedException(
                        "E_NEARBY_CONNECTIONS_REJECT_CONNECTION_FAILED",
                        "Failed to reject connection",
                        it
                    )
                )
            }
        }

        AsyncFunction("disconnect") { targetPeerId: String ->
            return@AsyncFunction nearbyConnectionsModule.disconnect(targetPeerId)
        }

        AsyncFunction("sendText") { targetPeerId: String, text: String, promise: Promise ->
            nearbyConnectionsModule.sendText(targetPeerId, text).addOnSuccessListener {
                promise.resolve(null)
            }.addOnFailureListener {
                promise.reject(
                    CodedException(
                        "E_NEARBY_CONNECTIONS_SEND_TEXT_FAILED", "Failed to send text", it
                    )
                )
            }
        }
    }

    private val nearbyConnectionsModule: NearbyConnectionsModule by lazy {
        val callbacks = NearbyConnectionCallbacks(
            onPeerFound = ::onPeerFound,
            onPeerLost = ::onPeerLost,
            onInvitationReceived = ::onInvitationReceived,
            onConnected = ::onConnected,
            onDisconnected = ::onDisconnected,
            onTextReceived = ::onTextReceived
        )
        NearbyConnectionsModule(context, callbacks)
    }

    private val context: Context
        get() = requireNotNull(appContext.reactContext) {
            "React Application Context is null"
        }

    private fun onPeerFound(peerId: String, name: String) {
        this@ExpoNearbyConnectionsModule.sendEvent(
            EventName.ON_PEER_FOUND.toString(), bundleOf(
                "peerId" to peerId,
                "name" to name,
            )
        )
    }

    private fun onPeerLost(peerId: String) {
        this@ExpoNearbyConnectionsModule.sendEvent(
            EventName.ON_PEER_LOST.toString(), bundleOf(
                "peerId" to peerId
            )
        )
    }

    private fun onInvitationReceived(peerId: String, name: String) {
        this@ExpoNearbyConnectionsModule.sendEvent(
            EventName.ON_INVITATION_RECEIVED.toString(), bundleOf(
                "peerId" to peerId,
                "name" to name
            )
        )
    }

    private fun onConnected(peerId: String, name: String) {
        this@ExpoNearbyConnectionsModule.sendEvent(
            EventName.ON_CONNECTED.toString(), bundleOf(
                "peerId" to peerId,
                "name" to name
            )
        )
    }

    private fun onDisconnected(peerId: String) {
        this@ExpoNearbyConnectionsModule.sendEvent(
            EventName.ON_DISCONNECTED.toString(), bundleOf(
                "peerId" to peerId
            )
        )
    }

    private fun onTextReceived(peerId: String, text: String) {
        this@ExpoNearbyConnectionsModule.sendEvent(
            EventName.ON_TEXT_RECEIVED.toString(), bundleOf(
                "peerId" to peerId,
                "text" to text
            )
        )
    }
}

File: ios/ExpoNearbyConnectionsModule.swift
import ExpoModulesCore
import MultipeerConnectivity

public class ExpoNearbyConnectionsModule: Module {
    private var nearbyConnection: MultipeerConnectivityModule = MultipeerConnectivityModule()
    
    public required init(appContext: AppContext) {
        super.init(appContext: appContext)
        self.nearbyConnection.delegate = self
    }
    
    public func definition() -> ModuleDefinition {
        Events(EventNames.allCases.map { $0.rawValue })
        
        Name(MODULE_NAME)
        
        AsyncFunction("isPlayServicesAvailable") { () -> Bool in
            return true
        }
        
        AsyncFunction("startAdvertise") { (name: String) -> String in
            let myPeerId = self.nearbyConnection.startAdvertise(name)
            
            return myPeerId
        }
        
        AsyncFunction("stopAdvertise") { () -> Void in
            return nearbyConnection.stopAdvertise()
        }
        
        AsyncFunction("startDiscovery") { (name: String) -> String in
            let myPeerId = nearbyConnection.startDiscovery(name)
            
            return myPeerId
        }
        
        AsyncFunction("stopDiscovery") { () -> Void in
            return nearbyConnection.stopDiscovery()
        }
        
        AsyncFunction("requestConnection") { (advertisePeerId: String, promise: Promise) -> Void in
            do {
                try nearbyConnection.requestConnection(to: advertisePeerId)
                promise.resolve(nil)
            } catch {
                promise.reject(error)
            }
        }
        
        AsyncFunction("acceptConnection") { (targetPeerId: String, promise: Promise) -> Void in
            do {
                try nearbyConnection.acceptConnection(to: targetPeerId)
                promise.resolve(nil)
            } catch {
                promise.reject(error)
            }
        }
        
        AsyncFunction("rejectConnection") { (targetPeerId: String, promise: Promise) -> Void in
            do {
                try nearbyConnection.rejectConnection(to: targetPeerId)
                promise.resolve(nil)
            } catch {
                promise.reject(error)
            }
        }
        
        AsyncFunction("disconnect") { () -> Void in
            return nearbyConnection.disconnect()
        }
        
        AsyncFunction("sendText") { (targetPeerId: String, text: String, promise: Promise) -> Void in
            do {
                try nearbyConnection.sendText(to: targetPeerId, payload: text)
                promise.resolve(nil)
            } catch {
                promise.reject(error)
            }
        }
    }
}

extension ExpoNearbyConnectionsModule: NearbyConnectionCallbackDelegate {
    func sendEvent(_ eventName: EventNames, _ body: [String: Any?] = [:]) {
        sendEvent(eventName.rawValue, body)
    }
    
    func onPeerFound(fromPeerId peerId: String, fromPeerName name: String) {
        sendEvent(.ON_PEER_FOUND, [
            "peerId": peerId,
            "name": name
        ])
    }
    
    func onPeerLost(fromPeerId peerId: String) {
        sendEvent(.ON_PEER_LOST, [
            "peerId": peerId
        ])
    }
    
    func onInvitationReceived(fromPeerId peerId: String, fromPeerName name: String) {
        sendEvent(.ON_INVITATION_RECIEVED, [
            "peerId": peerId,
            "name": name
        ])
    }
    
    func onConnected(fromPeerId peerId: String, fromPeerName name: String) {
        sendEvent(.ON_CONNECTED, [
            "peerId": peerId,
            "name": name
        ])
    }
    
    func onDisconnected(fromPeerId peerId: String) {
        sendEvent(.ON_DISCONNECTED, [
            "peerId": peerId
        ])
    }
    
    func onTextReceived(fromPeerId peerId: String, payload text: String) {
        sendEvent(.ON_TEXT_RECEIVED, [
            "peerId": peerId,
            "text": text
        ])
    }
}
